File: leetcode/letter_combinations_of_a_phone_number.js
/* Given a digit string, return all possible letter combinations that the 
 * number could represent.
 * Time Complexity: O(3^l)
 * Space Complexity: O(3^l), l is the length of the input
 */

/**
 * @param {string} digits
 * @return {string[]}
 */
var letterCombinations = function(digits) {
    if (digits.length === 0)
        return [];
    var map = ["0", "1", "abc", "def", "ghi", "jkl",
               "mno", "pqrs", "tuv", "wxyz"];
    var result = [[]];
    for (var digit of digits.split("")) {
        var temp = [];
        for (var r of result) {
            for (var d of map[parseInt(digit)].split(""))
                temp.push(r.concat([d]));
        }
        result = temp;
    }
    temp = [];
    for (var item of result)
        temp.push(item.join(""));
    return temp;
};

File: lintcode/letter_combinations_of_a_phone_number.py
class Solution:
    # @param {string} digits A digital string
    # @return {string[]} all posible letter combinations
    def letterCombinations(self, digits):
        if not digits: return []
        map = ["0", "1", "abc", "def", "ghi", "jkl",
               "mno", "pqrs", "tuv", "wxyz"]
        result = [[]]
        for digit in digits:
            result = [r + [d] for r in result for d in map[int(digit)]]
        return [''.join(item) for item in result]
